lib/srv/discovery/discovery_test.go

```diff
@@ -3231,15 +3231,17 @@ func TestAzureVMDiscovery(t *testing.T) {
 			emitter.t = t
 
 			if tc.discoveryConfig != nil {
+				sub := server.newDiscoveryConfigChangedSub()
+
 				_, err := tlsServer.Auth().DiscoveryConfigs.CreateDiscoveryConfig(ctx, tc.discoveryConfig)
 				require.NoError(t, err)
 
-				// Wait for the DiscoveryConfig to be added to the dynamic matchers
-				require.Eventually(t, func() bool {
-					server.muDynamicServerAzureFetchers.RLock()
-					defer server.muDynamicServerAzureFetchers.RUnlock()
-					return len(server.dynamicServerAzureFetchers) > 0
-				}, 1*time.Second, 50*time.Millisecond)
+				// wait for discovery config update
+				select {
+				case <-sub:
+				case <-time.After(3 * time.Second):
+					require.Fail(t, "timed out waiting for an update")
+				}
```

Comment:
```suggestion
				select {
				case <-sub:
				case <-time.After(3 * time.Second):
					require.Fail(t, "timed out waiting for an update")
				case <-t.Context().Done():
					require.Fail(t, "test context done while waiting for an update")
				}
```

Provide a quick way for the runtime to exit in case the test is aborted or exceeds the overall test deadline?